haproxy/logfile.py

```diff
@@ -336,48 +336,50 @@ def cmd_requests_per_minute(self):
 
         .. note::
           Try to combine it with time constrains (``-s`` and ``-d``) as this
-          command output can be huge otherwise.
+          command output can be huge otherwise. We can also use
+          ``requests_per_hour`` if it's too much data.
         """
-        if len(self._valid_lines) == 0:
-            return
-
-        current_minute = self._valid_lines[0].accept_date
-        current_minute_counter = 0
-        requests = []
-        one_minute = timedelta(minutes=1)
-
-        def format_and_append(append_to, date, counter):
-            seconds_and_micro = timedelta(
-                seconds=date.second, microseconds=date.microsecond
-            )
-            minute_formatted = date - seconds_and_micro
-            append_to.append((minute_formatted, counter))
-
-        # note that _valid_lines is kept sorted by date
-        for line in self._valid_lines:
-            line_date = line.accept_date
-            if (
-                line_date - current_minute < one_minute
-                and line_date.minute == current_minute.minute
-            ):
-                current_minute_counter += 1
-
-            else:
-                format_and_append(requests, current_minute, current_minute_counter)
-                current_minute_counter = 1
-                current_minute = line_date
-
-        if current_minute_counter > 0:
-            format_and_append(requests, current_minute, current_minute_counter)
+        return self._count_by_time(1)
 
-        return requests
+    def cmd_requests_per_hour(self):
+        """Generates statistics on how many requests were made per hour.
+        """
+        return self._count_by_time(60)
 
     def cmd_print(self):
         """Returns the raw lines to be printed."""
         if not self._valid_lines:
             return ''
         return '\n'.join([line.raw_line for line in self._valid_lines]) + '\n'
 
+    def _count_by_time(self, aggregate):
+        """
+        Helper function to compile stats by time frame.
+        Used by requests_per_minute and requests_per_hour
+        params:
+          aggregate: Number of minutes to aggregate on
+        """
+        if len(self._valid_lines) == 0:
+            return
+
+        requests = {}
+
+        # note that _valid_lines is kept sorted by date
+        for line in self._valid_lines:
+            tm = line.accept_date
+            tm += timedelta(minutes=aggregate / 2)
+            tm -= timedelta(
+                minutes=tm.minute % aggregate,
+                seconds=tm.second,
+                microseconds=tm.microsecond,
+            )
```

Comment:
@valleedelisle would you mind explaining the logic of this adding and removing? :thinking: I get the seconds and microseconds removal, as to zero them, but the `aggregate` puzzles me a little bit.

I was playing with the `request_per_minute.log` file and I see that what counts as a request that belongs to a minute goes from the half a minute before to the half a minute after, i.e.

Requests logged starting from `12:12:30` until `12:13:29` will all belong to `12:13:0` minute. Is what you expected to do?

At least, we could mention that it is not bound to the exact minute/hour to avoid users getting confused?

Reply:
@gforcada I based this part from this [stackoverflow question](https://stackoverflow.com/questions/3463930/how-to-round-the-minute-of-a-datetime-object-python). You're right when you say that `:30` to `:29` are going to be in the same minute. I didn't realize it would change the actual behavior maybe it would be better to keep the same behavior. The reason why I picked this is because we can pass any number of minutes to round up/down to.

Reply:
No actually it is fine, usually when I found solutions via stackoverflow I put a comment on top with the link to it, so one does not need lengthy explanations :+1: